### DISCUSSION TITLE
Maybe use the inbuilt `Mail Merge` feature.

### DISCUSSION THREAD

**Discussion by sjackson0109** [Set-OutlookSignatures/Set-OutlookSignatures#100] [Ideas]
We all know Outlook uses MS Word
And your CURRENTUSER, or CURRENTMANAGER variables are processed as string substitutions, using find/replace operations.

I feel there might be a better way, that could mean a fair bit of refactoring (if you are interested).

Quite some years ago, i wrote a VBScript version of the same kind of software [here](https://github.com/sjackson0109/signatures), reading from on-prem Active Directory the 'currently logged on user' and their manager attribute etc. It's got some problems, being mainly it's can only be executed by on-prem users, or VPN users only.

Handled all the Delegate Mailboxes/Additional Signatures, and Stationary and a few other bits.. much like your software. Mine went stale a large number of years back.. simply because the world moved away from VBSscripting

My version processed only HTML, RTF and TXT files, all wrapped in XSL Stylesheets (XMLDOM templates):
```
<?xml version='1.0' encoding='utf-8'?>
<xsl:stylesheet version="1.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform" xmlns:rs="urn:schemas-microsoft-com:rowset" xmlns:z="#RowsetSchema" exclude-result-prefixes="rs z">
<xsl:output method = "text" indent = "yes"/>
<xsl:template match="/">
<![CDATA[<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
		<p>]]><xsl:value-of select="xml/rs:data/z:row/@displayname"/><![CDATA[</p>

]]>
</xsl:template>

</xsl:stylesheet> 
```

Of course i'd connected to AD to get an record.. but I also looped through files (templates) and processed them all.

```
For Each oFile In oFolder.Files
   If not oFile.Name = "Thumbs.db" Then
      Call XSLTransform(record, oFile, oFile.Name)
   End If
Next 
```

Where all the hard work happened: XMLDOM (the mail-merge library):
```
Function XSLTransform(record, objFile, tFileName)
	'msgbox "Processing file: " & vbCrLf & "objFile.Name: " & objFile.Name & vbCrLf & "tFileName: " & tFileName, vbokonly, "Signature Rollout - ProcessXSL"
	Set xml = CreateObject("Microsoft.XMLDOM")
		record.Save xml, adPersistXML
		
		If NOT (tFileName = objFile.Name or tFileName = "") then
			'Process filenames as they their new name suggests
			Set xsl = CreateObject("Microsoft.XMLDOM")
			xsl.async = False
			'Load the XSL file. in the same path as the script, but it could be stored anywhere. e.g. Webserver
 			xsl.load objFile.Path 
			xsl.preserveWhiteSpace = preserveWhiteSpaceing
			If InStr(objFile.Name,".htm") then ext = ".htm" End If
			If InStr(objFile.Name,".rtf") then ext = ".rtf" End If
			If InStr(objFile.Name,".txt") then ext = ".txt" End If
			sOutputfile = tSignaturePath & tFileName & ext
			'msgbox "objFile: " & objFile & vbCrLf & "sOutputfile: " &  sOutputfile, vbOkOnly, "Signature Rollout - ProcessXSL PA"
			Set fso = CreateObject("Scripting.Filesystemobject")
				Set file = fso.OpenTextFile(sOutputFile,ForWriting,True)
					file.Write xml.transformNode(xsl)
				Set file = Nothing
			Set fso = Nothing
			Set xsl = Nothing	
		Else
			'Process filenames as the template had them
			Set xsl = CreateObject("Microsoft.XMLDOM")
			xsl.async = False
			'Load the XSL file. in the same path as the script, but it could be stored anywhere. e.g. Webserver
 			xsl.load objFile.Path 
			xsl.preserveWhiteSpace = preserveWhiteSpaceing
			sOutputfile = tSignaturePath & Replace(objFile.Name,".xsl","")
			'msgbox "objFile: " & objFile & vbCrLf & "sOutputfile: " & sOutputfile, vbOkOnly, "Signature Rollout - ProcessXSL NORM"
			Set fso = CreateObject("Scripting.Filesystemobject")
				Set file = fso.OpenTextFile(sOutputFile,ForWriting,True)
					file.Write xml.transformNode(xsl)
				Set file = Nothing
			Set fso = Nothing
			Set xsl = Nothing
		End If
	Set xml = Nothing
End Function
```

Why suggest this method?

It's cleaner to add additional conditional logic - such as `only show this cell/row if the mobile number already exists:
```
<xsl:if test="xml/rs:data/z:row/@mobile">
  <![CDATA[<TR style="height:24px">
     <TD style="WIDTH: 100%" colSpan=6>]]><xsl:value-of select="xml/rs:data/z:row/@mobile"/><![CDATA[</td>
  </TR>
]]></xsl:if>
```

or another, that am kinda proud of, format the email address field in lowercase, and colour it black, with an underline (not the default hyperlink style):
```
<xsl:if test="xml/rs:data/z:row/@mail">
   <![CDATA[<TR style="height:24px">
      <TD style="WIDTH: 100%" colSpan=6><a style="text-decoration:underline;color:black;" href="mailto:]]><xsl:value-of select="translate(xml/rs:data/z:row/@mail, 'ABCDEFGHIJKLMNOPQRSTUVWXYZÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖØÙÚÛÜÝÞŸŽŠŒ', 'abcdefghijklmnopqrstuvwxyzàáâãäåæçèéêëìíîïðñòóôõöøùúûüýþÿžšœ')" /><![CDATA[" target="_new">]]><xsl:value-of select="translate(xml/rs:data/z:row/@mail, 'ABCDEFGHIJKLMNOPQRSTUVWXYZÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖØÙÚÛÜÝÞŸŽŠŒ', 'abcdefghijklmnopqrstuvwxyzàáâãäåæçèéêëìíîïðñòóôõöøùúûüýþÿžšœ')" /><![CDATA[</a></TD>
   </TR>
]]></xsl:if>
```

You can even do Teams Chat hyperlinks:
```
<xsl:if test="xml/rs:data/z:row/@mail">
   <![CDATA[<a style="text-decoration:underline;color:black;" href="https://teams.microsoft.com/l/chat/0/0?users=]]><xsl:value-of select="translate(xml/rs:data/z:row/@mail, 'ABCDEFGHIJKLMNOPQRSTUVWXYZÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖØÙÚÛÜÝÞŸŽŠŒ', 'abcdefghijklmnopqrstuvwxyzàáâãäåæçèéêëìíîïðñòóôõöøùúûüýþÿžšœ')" />
     <![CDATA[" target="_new">Chat via Teams</a><br>
   ]]>
</xsl:if>
```

Hope this at least sparks some inspiration.

**Comment by GruberMarkus** [comment]
Hi Simon,

thank you very much for your feedback! I will think if and how I can incorporate some of your techniques to improve Set-OutlookSignatures.

Some first thoughts after going through your suggestions:
- Word is only used when DOCX templates or RTF signatures are involved. As much as possible is handled without Word automation, to keep the code reusable between screnarios, and to speed up operations. For example, search/replace in Word is speed up by first checking the text representation in PowerShell for existence of each variable to replace (what an irony, that performing more tasks is faster), and files (such as hyperlinks) are treated in a similar way.
- XMLDOM for sure is faster than Word automation and very flexible, but has the disadvantage of being complicated for the template maintainer. Set-OutlookSignatures has the goal to keep IT out of signature management where possible, and to make it easy for the Marketing or Communications department to maintain and change signatures (that's why DOCX tempates are popular, and are being used by around 80 % of known Set-OutlookSignatures clients).
- Only show a variable when it has a value: That's built-in, with some more options described in the `How to avoid blank lines when replacement variables return an empty string?` FAQ.
- Format email address in lower case, black and with an underline: You can do this, too:
  - Make the email address lower case by using a custom replacement variable config file: `$ReplaceHash['$CurrentUserMail$'] = [string]$ADPropsCurrentUser.mail.ToLower()`
  - Format the email address according to your need in the template file
- Hyperlinks are no problem, too. `tel:` and `mailto:` are easy, just search the README file for examples, or have a look at the sample templates. Using the same technique, you can construct links to start a Teams chat, or initiate a SIP call

**Comment by sjackson0109** [comment]
Bullet 1. Love it, thats the entire reason i would recomemnd NOT using/supporting DOCX file-types. HTML/TXT is sufficient. Who sends RTF these days?  There is a html2rtf parser library right? I guess the reason is convenience for the person creating the template. Which brings me onto your next point.

Bullet 2. It's already too complex for non-it staff to interpret. Don't get me wrong, it's not rocket science, but i guarantee, with all the marketing people i've worked with in life, who want to stamp corporate branding on everything.. they do a terrible job.  Don't understand image types; pixel depth, aspect ratio, dimensions (maybe some get this). Transparancy.. that utterly blew the mind of one person i know. 
Rather than say 'keep IT out of the equation' - why not provide a second tool, a WISIWIG editor of sorts, much like CODE TWO or EXCLAIMER. 
Infact the HTML templates I made years back can port into Exclaimer, with only a little bit of manipulation.  Not sure about Code Two. 

Bullet 3.. Additional variables to add extra formatting.  This only works if ALL the templates need that extra formatting (say `Tel: $CURRENTUSERTELEPHONE`). What if a receptionist wants something else (`Switchboard: $CURRENTUSERTELEPHONE`), and the rest of the business need it the other way.  I don't see how your additional variables can handle this.

Bullet 4: lowercase mail.. glad PS had made that sooo much easier without involving regex.  I recommend leaving it lowercase for the default configuration. 

Bullet 5: no real debate with you, except: `it's easy for IT staff` and `not so easy for the non-technical staff`. 


More...
Inserting the signature inside a table gives better control over formatting, line-wrapping, auto-collapse lines if empty. 

Probably got another dozen or so things to bring to the table, but am going for a break now, feel rubbish today.

**Comment by GruberMarkus** [comment]
Luckily, there are many different ways to solve things.

While DOCX is the format preferred by users, Set-OutlookSignatures also support HTM templates.

It is quite flexible and absolutely possible to define different formats or contents for specific templates.  
The most extreme case I have seen so far: A customer has almost 10 banners and several differently formatted dynamic text components in a single template file. The company, department, language and group affiliation of the logged-in user and/or the mailbox currently being processed are taken into account when selecting the banner and text components. It could be solved in other ways, by using multiple templates for example, but I think it's always nice to have a choice.

And I totally agree with you on tables in signatures: When used correctly, they can make signature design much easier and more visually appealing. At [ExplicIT Consulting](https://explicitconsulting.at/open-source/set-outlooksignatures/), we use two tables for these very reasons.

A WYSIWYG editor comes up every now and then, but in fact the vast majority of clients uses Word for this, as they prefer to work with DOCX templates. Interestingly, a big part of the clients using HTM templates use them because they simplye use an old template, and the number of clients which have good general HTML knowledge combined with signature specific knowledge is very low.
